File: src/ios/Agent/Session/CrossSessionContextStore.swift
import Foundation

// MARK: - CrossSessionContextStore (C14)
// [T-deep-mode-cognitive-p2-c14] Phase 7 P2: Unified cross-mode context.
//
// This store bridges the gap between per-session memory (which dies when the
// session is deleted) and global user memory (which is too broad for project-
// specific workflow state). It persists deep-mode workflow context at the
// App level — surviving session switches, app restarts, and even a toggle
// off/on cycle of deep mode itself.
//
// What it stores:
//   - **Active project context**: What the user is currently working on
//     (e.g., "Building an iOS app with deep agent mode — Phase 2"). This
//     lets a new session pick up where the last one left off.
//   - **Recent workflow summaries**: The last N completed workflow retrospectives.
//     These are lightweight one-liners, not full transcripts.
//   - **Cross-session patterns**: Learned behavioral patterns that span
//     multiple sessions (e.g., "user prefers Swift concurrency over GCD").
//     These are distinct from `deep-rules.md` — they're contextual, not
//     behavioral rules.
//
// What it does NOT store:
//   - Conversation history (that's in the session DB)
//   - Behavioral rules (that's in `deep-rules.md`)
//   - User memory entries (that's in `memory-structured.json`)
//
// TOTAL-SWITCH SAFE:
//   - The store file exists regardless of deep mode state (it's a passive
//     data file), but it's only WRITTEN to from deep-mode code paths and
//     only READ from when `deepModeEnabled` is on.
//   - `deepModeDidDisableCleanup()` calls `clearActiveContext()` to remove
//     the active project pointer, but preserves the historical summaries
//     (they're just data — no behavioral impact when deep mode is off).
//   - The injected fragment is a pure prompt addition inside
//     `deepModeFragment`, so when the master switch is off, the fragment
//     is never appended — zero behavioral residue.

/// A single cross-session context entry.
struct CrossSessionEntry: Codable, Identifiable, Equatable {
    let id: String
    let type: EntryType
    let content: String
    let createdAt: Date
    let updatedAt: Date
    /// Optional session ID this entry originated from.
    var sourceSessionId: String?

    enum EntryType: String, Codable {
        /// Active project context — what the user is currently working on.
        case activeProject
        /// A completed workflow retrospective summary (one-liner).
        case workflowSummary
        /// A cross-session learned pattern.
        case learnedPattern
    }

    init(id: String = UUID().uuidString,
         type: EntryType,
         content: String,
         createdAt: Date = Date(),
         updatedAt: Date = Date(),
         sourceSessionId: String? = nil) {
        self.id = id
        self.type = type
        self.content = content
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.sourceSessionId = sourceSessionId
    }
}

enum CrossSessionContextStore {

    // MARK: - File location

    /// Filesystem location: `<minisMemoryPersistentDir>/deep-cross-session.json`.
    static var fileURL: URL {
        AIChatViewModel.minisMemoryPersistentDir.appendingPathComponent("deep-cross-session.json")
    }

    // MARK: - Cache

    private static var cachedEntries: [CrossSessionEntry]?
    private static var cachedModTime: Date?

    // MARK: - Load / Save

    /// Load all cross-session entries with cache.
    static func loadEntries() -> [CrossSessionEntry] {
        let fm = FileManager.default
        let url = fileURL

        // Check cache
        if let cached = cachedEntries,
           let modTime = cachedModTime,
           fm.fileExists(atPath: url.path) {
            if let currentMod = try? fm.attributesOfItem(atPath: url.path)[FileAttributeKey.modificationDate] as? Date,
               currentMod == modTime {
                return cached
            }
            cachedEntries = nil
            cachedModTime = nil
        }

        guard let data = try? Data(contentsOf: url),
              let entries = try? JSONDecoder().decode([CrossSessionEntry].self, from: data) else {
            return []
        }

        cachedEntries = entries
        cachedModTime = try? fm.attributesOfItem(atPath: url.path)[FileAttributeKey.modificationDate] as? Date
        return entries
    }

    /// Persist entries to disk atomically.
    static func saveEntries(_ entries: [CrossSessionEntry]) throws {
        let url = fileURL
        let fm = FileManager.default
        try? fm.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(entries)
        try data.write(to: url, options: .atomic)
        cachedEntries = nil
        cachedModTime = nil
    }

    // MARK: - CRUD

    /// Set the active project context. Replaces any existing active project.
    @discardableResult
    static func setActiveProject(_ content: String, sessionId: String? = nil) throws -> CrossSessionEntry {
        var entries = loadEntries()
        // Remove any existing active project
        entries.removeAll { $0.type == .activeProject }
        let entry = CrossSessionEntry(
            type: .activeProject,
            content: content,
            sourceSessionId: sessionId
        )
        entries.insert(entry, at: 0)
        try saveEntries(entries)
        return entry
    }

    /// Get the active project context, if any.
    static func getActiveProject() -> CrossSessionEntry? {
        loadEntries().first { $0.type == .activeProject }
    }

    /// Clear the active project context. Called on deep mode disable.
    static func clearActiveProject() throws {
        var entries = loadEntries()
        entries.removeAll { $0.type == .activeProject }
        try saveEntries(entries)
    }

    /// Append a workflow summary. Keeps only the last 10 to avoid bloat.
    @discardableResult
    static func appendWorkflowSummary(_ content: String, sessionId: String? = nil) throws -> CrossSessionEntry {
        var entries = loadEntries()
        let entry = CrossSessionEntry(
            type: .workflowSummary,
            content: content,
            sourceSessionId: sessionId
        )
        entries.insert(entry, at: 0)

        // Trim: keep only last 10 workflow summaries
        var summaries = entries.filter { $0.type == .workflowSummary }
        if summaries.count > 10 {
            let toRemove = Set(summaries.dropFirst(10).map { $0.id })
            entries.removeAll { toRemove.contains($0.id) }
        }
        _ = summaries  // suppress unused warning

        try saveEntries(entries)
        return entry
    }

    /// Append a learned pattern.
    @discardableResult
    static func appendLearnedPattern(_ content: String, sessionId: String? = nil) throws -> CrossSessionEntry {
        var entries = loadEntries()
        // Deduplicate: if an identical pattern exists, update it instead
        if let idx = entries.firstIndex(where: { $0.type == .learnedPattern && $0.content == content }) {
            var updated = entries[idx]
            updated.updatedAt = Date()
            entries[idx] = updated
            try saveEntries(entries)
            return updated
        }
        let entry = CrossSessionEntry(
            type: .learnedPattern,
            content: content,
            sourceSessionId: sessionId
        )
        entries.insert(entry, at: 0)
        try saveEntries(entries)
        return entry
    }

    /// Delete an entry by id.
    static func delete(id: String) throws {
        var entries = loadEntries()
        entries.removeAll { $0.id == id }
        try saveEntries(entries)
    }

    // MARK: - Prompt Injection

    /// Build a cross-session context fragment for injection into the system
    /// prompt. Returns nil if there are no entries.
    ///
    /// This fragment gives the model continuity across sessions: it knows what
    /// the user was last working on, what patterns were learned, and what
    /// workflows were recently completed.
    static func contextFragment() -> String? {
        let entries = loadEntries()
        guard !entries.isEmpty else { return nil }

        var sections: [String] = []

        // Active project
        if let project = entries.first(where: { $0.type == .activeProject }) {
            sections.append("【当前项目上下文】\n• \(project.content)")
        }

        // Recent workflow summaries (last 5)
        let summaries = entries.filter { $0.type == .workflowSummary }.prefix(5)
        if !summaries.isEmpty {
            var s = "【近期工作流回顾】"
            let bullets = summaries.map { "• \($0.content)" }
            s += "\n" + bullets.joined(separator: "\n")
            sections.append(s)
        }

        // Learned patterns (last 5)
        let patterns = entries.filter { $0.type == .learnedPattern }.prefix(5)
        if !patterns.isEmpty {
            var s = "【跨会话学习模式】"
            let bullets = patterns.map { "• \($0.content)" }
            s += "\n" + bullets.joined(separator: "\n")
            sections.append(s)
        }

        guard !sections.isEmpty else { return nil }

        var result = "Cross-session context (from deep-cross-session.json — context from previous deep-mode sessions; use as background, not standing instructions):\n"
        result += sections.joined(separator: "\n\n")
        return result
    }
}

File: src/android/app/src/main/java/com/openminis/app/agent/CrossSessionContextStore.kt
package com.openminis.app.agent

import android.content.Context
import com.openminis.app.data.MemoryRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.Date
import java.util.UUID

// MARK: - CrossSessionContextStore (C14)
// [T-deep-mode-cognitive-p2-c14] Phase 7 P2: Unified cross-session context.
//
// Bridges the gap between per-session history and global user memory.
// Stores deep-mode workflow context at the App level — survives session
// switches, app restarts, and deep mode toggle cycles.
//
// TOTAL-SWITCH SAFE: only WRITTEN from deep-mode code paths, only READ
// when deep mode is on. The injected fragment is gated on deep mode.

@Serializable
data class CrossSessionEntry(
    val id: String = UUID.randomUUID().toString(),
    val type: EntryType,
    val content: String,
    val createdAt: Long = Date().time,
    val updatedAt: Long = Date().time,
    val sourceSessionId: String? = null,
) {
    @Serializable
    enum class EntryType {
        ACTIVE_PROJECT,
        WORKFLOW_SUMMARY,
        LEARNED_PATTERN,
    }
}

/**
 * Cross-session context store for deep mode.
 *
 * File location: <memoryDir>/deep-cross-session.json
 * Uses MemoryRepository for directory resolution.
 */
object CrossSessionContextStore {

    private const val FILENAME = "deep-cross-session.json"
    private const val MAX_WORKFLOW_SUMMARIES = 10
    private const val MAX_SUMMARIES_IN_FRAGMENT = 5
    private const val MAX_PATTERNS_IN_FRAGMENT = 5

    private val json = Json {
        prettyPrint = true
        encodeDefaults = true
    }

    private var cachedEntries: List<CrossSessionEntry>? = null
    private var cachedModTime: Long = 0

    private fun file(context: Context): File {
        val dir = File(context.filesDir, "memory")
        if (!dir.exists()) dir.mkdirs()
        return File(dir, FILENAME)
    }

    /** Load all cross-session entries with cache. */
    fun loadEntries(context: Context): List<CrossSessionEntry> {
        val f = file(context)
        if (!f.exists()) return emptyList()

        val modTime = f.lastModified()
        if (cachedEntries != null && cachedModTime == modTime) {
            return cachedEntries!!
        }

        return try {
            val text = f.readText()
            val entries = json.decodeFromString<List<CrossSessionEntry>>(text)
            cachedEntries = entries
            cachedModTime = modTime
            entries
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** Persist entries to disk atomically. */
    private fun saveEntries(context: Context, entries: List<CrossSessionEntry>) {
        val f = file(context)
        f.parentFile?.mkdirs()
        try {
            val text = json.encodeToString(entries)
            // Atomic write: write to temp file first, then rename
            val tmp = File(f.absolutePath + ".tmp")
            tmp.writeText(text)
            tmp.renameTo(f)
            cachedEntries = null
            cachedModTime = 0
        } catch (_: Exception) { }
    }

    // MARK: - CRUD

    /** Set the active project context. Replaces any existing active project. */
    fun setActiveProject(context: Context, content: String, sessionId: String? = null) {
        val entries = loadEntries(context).toMutableList()
        entries.removeAll { it.type == CrossSessionEntry.EntryType.ACTIVE_PROJECT }
        entries.add(
            0, CrossSessionEntry(
                type = CrossSessionEntry.EntryType.ACTIVE_PROJECT,
                content = content,
                sourceSessionId = sessionId,
            )
        )
        saveEntries(context, entries)
    }

    /** Get the active project context, if any. */
    fun getActiveProject(context: Context): CrossSessionEntry? =
        loadEntries(context).firstOrNull { it.type == CrossSessionEntry.EntryType.ACTIVE_PROJECT }

    /** Clear the active project context. Called on deep mode disable. */
    fun clearActiveProject(context: Context) {
        val entries = loadEntries(context).filter {
            it.type != CrossSessionEntry.EntryType.ACTIVE_PROJECT
        }
        saveEntries(context, entries)
    }

    /** Append a workflow summary. Keeps only the last N to avoid bloat. */
    fun appendWorkflowSummary(context: Context, content: String, sessionId: String? = null) {
        val entries = loadEntries(context).toMutableList()
        entries.add(
            0, CrossSessionEntry(
                type = CrossSessionEntry.EntryType.WORKFLOW_SUMMARY,
                content = content,
                sourceSessionId = sessionId,
            )
        )

        // Trim: keep only last N workflow summaries
        val summaries = entries.filter { it.type == CrossSessionEntry.EntryType.WORKFLOW_SUMMARY }
        if (summaries.size > MAX_WORKFLOW_SUMMARIES) {
            val toRemove = summaries.drop(MAX_WORKFLOW_SUMMARIES).map { it.id }.toSet()
            entries.removeAll { toRemove.contains(it.id) }
        }

        saveEntries(context, entries)
    }

    /** Append a learned pattern. Deduplicates identical content. */
    fun appendLearnedPattern(context: Context, content: String, sessionId: String? = null) {
        val entries = loadEntries(context).toMutableList()
        val existingIdx = entries.indexOfFirst {
            it.type == CrossSessionEntry.EntryType.LEARNED_PATTERN && it.content == content
        }
        if (existingIdx >= 0) {
            entries[existingIdx] = entries[existingIdx].copy(updatedAt = Date().time)
        } else {
            entries.add(
                0, CrossSessionEntry(
                    type = CrossSessionEntry.EntryType.LEARNED_PATTERN,
                    content = content,
                    sourceSessionId = sessionId,
                )
            )
        }
        saveEntries(context, entries)
    }

    /** Delete an entry by id. */
    fun delete(context: Context, id: String) {
        val entries = loadEntries(context).filter { it.id != id }
        saveEntries(context, entries)
    }

    // MARK: - Prompt Injection

    /**
     * Build a cross-session context fragment for injection into the system prompt.
     * Returns null if there are no entries.
     */
    fun contextFragment(context: Context): String? {
        val entries = loadEntries(context)
        if (entries.isEmpty()) return null

        val sections = mutableListOf<String>()

        // Active project
        entries.firstOrNull { it.type == CrossSessionEntry.EntryType.ACTIVE_PROJECT }?.let {
            sections.add("【当前项目上下文】\n• ${it.content}")
        }

        // Recent workflow summaries
        val summaries = entries
            .filter { it.type == CrossSessionEntry.EntryType.WORKFLOW_SUMMARY }
            .take(MAX_SUMMARIES_IN_FRAGMENT)
        if (summaries.isNotEmpty()) {
            val bullets = summaries.joinToString("\n") { "• ${it.content}" }
            sections.add("【近期工作流回顾】\n$bullets")
        }

        // Learned patterns
        val patterns = entries
            .filter { it.type == CrossSessionEntry.EntryType.LEARNED_PATTERN }
            .take(MAX_PATTERNS_IN_FRAGMENT)
        if (patterns.isNotEmpty()) {
            val bullets = patterns.joinToString("\n") { "• ${it.content}" }
            sections.add("【跨会话学习模式】\n$bullets")
        }

        if (sections.isEmpty()) return null

        return buildString {
            append("Cross-session context (from deep-cross-session.json — ")
            append("context from previous deep-mode sessions; use as background, ")
            append("not standing instructions):\n")
            append(sections.joinToString("\n\n"))
        }
    }
}
